pennylane/ops/qubit/hamiltonian.py

```diff
@@ -764,3 +781,125 @@ def map_wires(self, wire_map: dict):
+        if not is_pauli:
+            obs = []
+            for o in qml.operation.Tensor(op).obs:
+                if len(o.wires) > 1:
+                    # todo: deal with operations created from multi-qubit operations such as Hermitian
+                    raise ValueError(
+                        f"Can only sparsify Hamiltonians whose constituent observables consist of "
+                        f"(tensor products of) single-qubit operators; got {op}."
+                    )
+                obs.append(o.matrix())
+
+            # Array to store the single-wire observables which will be Kronecker producted together
+            mat = []
+            # i_count tracks the number of consecutive single-wire identity matrices encountered
+            # in order to avoid unnecessary Kronecker products, since I_n x I_m = I_{n+m}
+            i_count = 0
+            apply_coeff = False
+            for wire_lab in wires:
+                if wire_lab in op.wires:
+                    if i_count > 0:
+                        mat.append(scipy.sparse.eye(2**i_count, format="coo"))
+                    i_count = 0
+                    idx = op.wires.index(wire_lab)
+                    # obs is an array storing the single-wire observables which
+                    # make up the full Hamiltonian term
+                    sp_obs = scipy.sparse.coo_matrix(obs[idx])
+                    if not apply_coeff:
+                        sp_obs = sp_obs * coeff
+                        apply_coeff = True
+                    mat.append(sp_obs)
+                else:
+                    i_count += 1
+
+            if i_count > 0:
+                mat.append(scipy.sparse.eye(2**i_count, format="coo"))
+
+            red_mat = functools.reduce(lambda i, j: scipy.sparse.kron(i, j, format="coo"), mat)
+        else:
+            red_mat = pauli_word_2_mat(pauli_indices[w, :], coeff=coeff)
+        if tmp_init:
+            tmp.data += red_mat.data
+        else:
+            tmp = red_mat.tocsr().astype(dtype="complex128")
+            tmp_init = True
+    return tmp
+
+
+@functools.lru_cache
+def cached_arange(n):
+    """Returns a cached arange for ``pauli_word_2_mat``."""
+    return np.arange(n, dtype=np.int64)
+
+
+@functools.lru_cache
+def cached_data0(ind):
+    """Returns the sparse data and indices of a Pauli matrix for ``pauli_word_2_mat``."""
+    if ind == 0:
+        data = np.array([1.0, 1.0], dtype=np.complex128)
+        indices = np.array([0, 1], dtype=np.int64)
+    elif ind == 1:
+        data = np.array([1.0, 1.0], dtype=np.complex128)
+        indices = np.array([1, 0], dtype=np.int64)
+    elif ind == 2:
+        data = np.array([-1.0j, 1.0j], dtype=np.complex128)
+        indices = np.array([1, 0], dtype=np.int64)
+    elif ind == 3:
+        data = np.array([1.0, -1.0], dtype=np.complex128)
+        indices = np.array([0, 1], dtype=np.int64)
+    return data, indices
+
+
+def pauli_word_2_mat(vec, coeff=1.0):
```

Comment:
Think we could add a version of this to [pauli_airthmetic.py](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pauli/pauli_arithmetic.py)?

We have `PauliWord.to_mat` which will construct either the dense or sparse matrices.  We could use this logic in that case.

Reply:
Yes, I was expecting someone to propose this. I found `pauli_word_to_matrix` in `utils` by it didn't seem used anywhere. I'll discuss with @BorjaRequena and see how to move it there.

Reply:
I have done a proposal to move this to the `PauliWord` class in #4272. Let me know what you think 😄